src/cluster/index.js

```diff
@@ -20,131 +18,94 @@ const mergeTopics = (obj, { topic, partitions }) => ({
+
 module.exports = class Cluster {
   /**
    * @param {Object} options
-   * @param {Array<string>} options.brokers example: ['127.0.0.1:9092', '127.0.0.1:9094']
-   * @param {Object} options.ssl
-   * @param {Object} options.sasl
-   * @param {string} options.clientId
-   * @param {number} options.connectionTimeout - in milliseconds
-   * @param {number} options.authenticationTimeout - in milliseconds
-   * @param {number} options.reauthenticationThreshold - in milliseconds
-   * @param {number} [options.requestTimeout=30000] - in milliseconds
-   * @param {number} options.metadataMaxAge - in milliseconds
-   * @param {boolean} options.allowAutoTopicCreation
-   * @param {number} options.maxInFlightRequests
    * @param {number} options.isolationLevel
-   * @param {Object} options.retry
+   * @param {number} [options.targetTopicsLockTimeout=30000] - in milliseconds
+   * @param {import("../../types").RetryOptions} options.retry
    * @param {import("../../types").Logger} options.logger
-   * @param {Map} options.offsets
-   * @param {import("../instrumentation/emitter")} [options.instrumentationEmitter=null]
+   * @param {Map} [options.offsets]
+   * @param {import("./brokerPool")} options.brokerPool
    */
   constructor({
     logger: rootLogger,
-    socketFactory,
-    brokers,
-    ssl,
-    sasl,
-    clientId,
-    connectionTimeout,
-    authenticationTimeout,
-    reauthenticationThreshold,
-    requestTimeout = 30000,
-    enforceRequestTimeout,
-    metadataMaxAge,
+    targetTopicsLockTimeout = 30000,
     retry,
-    allowAutoTopicCreation,
-    maxInFlightRequests,
     isolationLevel,
-    instrumentationEmitter = null,
     offsets = new Map(),
+    brokerPool,
   }) {
     this.rootLogger = rootLogger
     this.logger = rootLogger.namespace('Cluster')
-    this.retry = { ...retry }
-    this.retrier = createRetry(this.retry)
-    this.connectionBuilder = connectionBuilder({
-      logger: rootLogger,
-      instrumentationEmitter,
-      socketFactory,
-      brokers,
-      ssl,
-      sasl,
-      clientId,
-      connectionTimeout,
-      requestTimeout,
-      enforceRequestTimeout,
-      maxInFlightRequests,
-      retry,
-    })
+    this.retrier = createRetry(retry)
 
     this.targetTopics = new Set()
     this.mutatingTargetTopics = new Lock({
       description: `updating target topics`,
-      timeout: requestTimeout,
+      timeout: targetTopicsLockTimeout,
     })
     this.isolationLevel = isolationLevel
-    this.brokerPool = new BrokerPool({
-      connectionBuilder: this.connectionBuilder,
-      logger: this.rootLogger,
-      retry,
-      allowAutoTopicCreation,
-      authenticationTimeout,
-      reauthenticationThreshold,
-      metadataMaxAge,
-    })
+    this.brokerPool = brokerPool
+    this.clusterHandle = nextClusterHandle++
     this.committedOffsetsByGroup = offsets
   }
 
   isConnected() {
+    // XXX: Would this also mean "... and consider me one of the stakeholders?"
```

Comment:
This is a tricky one. I would look into where this is called from, and if we can maybe get rid of it in favor of an idempotent connect.